File: src/test/java/recipeapplication/service/UserServiceTest.java
package recipeapplication.service;

import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.security.crypto.password.PasswordEncoder;
import recipeapplication.dto.UserDto;
import recipeapplication.exception.InvalidPasswordTokenException;
import recipeapplication.exception.UserNotFoundException;
import recipeapplication.model.PasswordResetToken;
import recipeapplication.model.User;
import recipeapplication.repository.PasswordTokenRepository;
import recipeapplication.repository.UserRepository;
import recipeapplication.security.RecipeUserDetails;
import recipeapplication.security.Role;

import java.util.Calendar;
import java.util.Optional;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

public class UserServiceTest
{
    private static final String INVALID_USERNAME = "INVALID_USERNAME";
    private static final String INVALID_EMAIL = "INVALID_EMAIL";
    private static final String VALID_EMAIL = "VALID_EMAIL";
    private static final String VALID_USERNAME = "VALID_USERNAME";
    private static final String SERVER_NAME = "SERVER_NAME";
    private static final String VALID_TOKEN = "VALID_TOKEN";
    private static final String INVALID_TOKEN = "INVALID_TOKEN";
    private static final String EXPIRED_TOKEN = "EXPIRED_TOKEN";
    private static final String PASSWORD = "PASSWORD";
    private static final String ENCODED_PASSWORD = "ENCODED_PASSWORD";

    private UserRepository userRepository;
    private PasswordTokenRepository passwordTokenRepository;
    private PasswordEncoder passwordEncoder;
    private AuthService authService;
    private UserService userService;
    private EmailService emailService;

    @Before
    public void setup()
    {
        userRepository = mock(UserRepository.class);

        when(userRepository.findByUsername(INVALID_USERNAME)).thenReturn(Optional.empty());
        when(userRepository.findByEmail(INVALID_EMAIL)).thenReturn(Optional.empty());
        when(userRepository.findByUsername(VALID_USERNAME)).thenReturn(Optional.of(new User()));
        when(userRepository.findByEmail(VALID_EMAIL)).thenReturn(Optional.of(new User()));

        Calendar oneHourAgo = Calendar.getInstance();
        oneHourAgo.add(Calendar.HOUR_OF_DAY, -1);

        Calendar oneHourInTheFuture = Calendar.getInstance();
        oneHourInTheFuture.add(Calendar.HOUR_OF_DAY, 1);

        User user = new User();
        user.setUsername(VALID_USERNAME);

        PasswordResetToken validPasswordResetToken = new PasswordResetToken();
        validPasswordResetToken.setExpiryDate(oneHourInTheFuture.getTime());
        validPasswordResetToken.setUser(user);

        PasswordResetToken expiredPasswordResetToken = new PasswordResetToken();
        expiredPasswordResetToken.setExpiryDate(oneHourAgo.getTime());

        passwordTokenRepository = mock(PasswordTokenRepository.class);

        when(passwordTokenRepository.findByToken(INVALID_TOKEN)).thenReturn(Optional.empty());
        when(passwordTokenRepository.findByToken(VALID_TOKEN)).thenReturn(Optional.of(validPasswordResetToken));
        when(passwordTokenRepository.findByToken(EXPIRED_TOKEN)).thenReturn(Optional.of(expiredPasswordResetToken));

        passwordEncoder = mock(PasswordEncoder.class);

        when(passwordEncoder.encode(PASSWORD)).thenReturn(ENCODED_PASSWORD);

        authService = mock(AuthService.class);
        emailService = mock(EmailService.class);

        userService = new UserService(userRepository, passwordTokenRepository, emailService, passwordEncoder, authService);
    }

    @Test(expected = UserNotFoundException.class)
    public void shouldThrowUserNotFoundExceptionWhenNoUsernameOrEmailInRequest() throws UserNotFoundException
    {
        userService.createPasswordResetToken("", new UserDto());
    }

    @Test(expected = UserNotFoundException.class)
    public void shouldThrowUserNotFoundExceptionWhenUsernameDoesNotExist() throws UserNotFoundException
    {
        UserDto userDto = new UserDto();
        userDto.setUsername(INVALID_USERNAME);

        userService.createPasswordResetToken(SERVER_NAME, userDto);
    }

    @Test(expected = UserNotFoundException.class)
    public void shouldThrowUserNotFoundExceptionWhenEmailDoesNotExist() throws UserNotFoundException
    {
        UserDto userDto = new UserDto();
        userDto.setEmail(INVALID_EMAIL);

        userService.createPasswordResetToken(SERVER_NAME, userDto);
    }

    @Test
    public void shouldCreatePasswordResetTokenForValidUsername() throws UserNotFoundException
    {
        UserDto userDto = new UserDto();
        userDto.setUsername(VALID_USERNAME);

        ArgumentCaptor<PasswordResetToken> argumentCaptor = ArgumentCaptor.forClass(PasswordResetToken.class);

        userService.createPasswordResetToken(SERVER_NAME, userDto);

        verify(emailService).sendPasswordReset(any(User.class), anyString(), eq(SERVER_NAME));
        verify(passwordTokenRepository).save(argumentCaptor.capture());

        assertNotNull(argumentCaptor.getValue().getUser());
        assertNotNull(argumentCaptor.getValue().getToken());
    }

    @Test
    public void shouldCreatePasswordResetTokenForValidEmail() throws UserNotFoundException
    {
        UserDto userDto = new UserDto();
        userDto.setEmail(VALID_EMAIL);

        ArgumentCaptor<PasswordResetToken> argumentCaptor = ArgumentCaptor.forClass(PasswordResetToken.class);

        userService.createPasswordResetToken(SERVER_NAME, userDto);

        verify(emailService).sendPasswordReset(any(User.class), anyString(), eq(SERVER_NAME));
        verify(passwordTokenRepository).save(argumentCaptor.capture());

        assertNotNull(argumentCaptor.getValue().getUser());
        assertNotNull(argumentCaptor.getValue().getToken());
    }

    @Test(expected = InvalidPasswordTokenException.class)
    public void shouldThrowInvalidPasswordTokenExceptionWhenTokenDoesNotExist() throws InvalidPasswordTokenException
    {
        userService.processPasswordResetToken(INVALID_TOKEN);
    }

    @Test(expected = InvalidPasswordTokenException.class)
    public void shouldThrowInvalidPasswordTokenExceptionWhenTokenIsExpired() throws InvalidPasswordTokenException
    {
        userService.processPasswordResetToken(EXPIRED_TOKEN);
    }

    @Test
    public void shouldEnablePasswordResetForValidToken() throws InvalidPasswordTokenException
    {
        ArgumentCaptor<RecipeUserDetails> argumentCaptor = ArgumentCaptor.forClass(RecipeUserDetails.class);

        userService.processPasswordResetToken(VALID_TOKEN);

        verify(authService).authenticateUser(argumentCaptor.capture(), eq(Role.CHANGE_PASSWORD));

        RecipeUserDetails recipeUserDetails = argumentCaptor.getValue();

        assertEquals(VALID_USERNAME, recipeUserDetails.getUser().getUsername());
        assertTrue(recipeUserDetails.isChangePasswordAccess());
    }

    @Test
    public void shouldChangePasswordSuccessfully()
    {
        User user = new User();

        ArgumentCaptor<User> argumentCaptor = ArgumentCaptor.forClass(User.class);

        userService.changePassword(user, PASSWORD);

        verify(userRepository).save(argumentCaptor.capture());
        verify(passwordTokenRepository).deleteByUser(user);
        verify(authService).disablePasswordReset();

        assertEquals(ENCODED_PASSWORD, argumentCaptor.getValue().getPassword());
    }

    @Test(expected = UserNotFoundException.class)
    public void shouldThrowUserNotFoundExceptionWhenUserDoesNotExist() throws UserNotFoundException
    {
        when(userRepository.findByUsername(INVALID_USERNAME)).thenReturn(Optional.empty());

        userService.getUser(INVALID_USERNAME);
    }

    @Test
    public void shouldGetUserWhenValidUsernameProvided() throws UserNotFoundException
    {
        User user = new User();

        when(userRepository.findByUsername(VALID_USERNAME)).thenReturn(Optional.of(user));

        assertEquals(user, userService.getUser(VALID_USERNAME));
    }

    @Test
    public void shouldDeleteAccountSuccessfully()
    {
        User user = new User();

        when(authService.getLoggedInUser()).thenReturn(user);

        userService.deleteAccount();

        verify(userRepository).delete(user);
    }

    @Test
    public void shouldTurnOffInstructionsIfTheUserIsNew()
    {
        User user = new User();
        user.setNewUser(true);

        when(authService.getLoggedInUser()).thenReturn(user);

        userService.turnOffInstructions();

        ArgumentCaptor<User> argumentCaptor = ArgumentCaptor.forClass(User.class);

        verify(userRepository).save(argumentCaptor.capture());

        assertFalse(argumentCaptor.getValue().isNewUser());
    }

    @Test
    public void shouldNotTurnOffInstructionsIfTheUserIsNotNew()
    {
        User user = new User();
        user.setNewUser(false);

        when(authService.getLoggedInUser()).thenReturn(user);

        userService.turnOffInstructions();

        verify(userRepository, never()).save(any());
    }
}

File: src/test/kotlin/thecookingpot/recipe/service/UserServiceTest.kt
package thecookingpot.recipe.service

import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.junit.Before
import org.junit.Test
import org.springframework.security.crypto.password.PasswordEncoder
import thecookingpot.recipe.dto.PasswordResetDto
import thecookingpot.recipe.exception.InvalidPasswordTokenException
import thecookingpot.recipe.exception.UserNotFoundException
import thecookingpot.recipe.model.PasswordResetToken
import thecookingpot.recipe.model.User
import thecookingpot.recipe.repository.PasswordTokenRepository
import thecookingpot.recipe.repository.UserRepository
import thecookingpot.security.RecipeUserDetails
import thecookingpot.security.Role
import thecookingpot.security.service.AuthService
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class UserServiceTest {
    companion object {
        private const val INVALID_USERNAME = "INVALID_USERNAME"
        private const val INVALID_EMAIL = "INVALID_EMAIL"
        private const val VALID_EMAIL = "VALID_EMAIL"
        private const val VALID_USERNAME = "VALID_USERNAME"
        private const val SERVER_NAME = "SERVER_NAME"
        private const val VALID_TOKEN = "VALID_TOKEN"
        private const val INVALID_TOKEN = "INVALID_TOKEN"
        private const val EXPIRED_TOKEN = "EXPIRED_TOKEN"
        private const val PASSWORD = "PASSWORD"
        private const val ENCODED_PASSWORD = "ENCODED_PASSWORD"
    }

    private lateinit var userRepository: UserRepository
    private lateinit var passwordTokenRepository: PasswordTokenRepository
    private lateinit var passwordEncoder: PasswordEncoder
    private lateinit var authService: AuthService
    private lateinit var userService: UserService
    private lateinit var emailService: EmailService

    @Before
    fun setup() {
        userRepository = mockk(relaxed = true)
        passwordTokenRepository = mockk(relaxed = true)
        emailService = mockk(relaxed = true)
        passwordEncoder = mockk()
        authService = mockk(relaxed = true)

        every { userRepository.findByUsername(INVALID_USERNAME) } returns null
        every { userRepository.findByEmail(INVALID_EMAIL) } returns null
        every { userRepository.findByUsername(VALID_USERNAME) } returns User()
        every { userRepository.findByEmail(VALID_EMAIL) } returns User()

        val user = User().apply { username = VALID_USERNAME }
        val oneHourAgo = Calendar.getInstance().apply { add(Calendar.HOUR_OF_DAY, -1) }
        val oneHourInTheFuture = Calendar.getInstance().apply { add(Calendar.HOUR_OF_DAY, 1) }

        val validPasswordResetToken = PasswordResetToken().apply {
            setExpiryDate(oneHourInTheFuture.time)
            this.user = user
        }

        val expiredPasswordResetToken = PasswordResetToken().apply {
            setExpiryDate(oneHourAgo.time)
        }

        every { passwordTokenRepository.findByToken(INVALID_TOKEN) } returns Optional.empty()
        every { passwordTokenRepository.findByToken(VALID_TOKEN) } returns Optional.of(validPasswordResetToken)
        every { passwordTokenRepository.findByToken(EXPIRED_TOKEN) } returns Optional.of(expiredPasswordResetToken)

        every { passwordEncoder.encode(PASSWORD) } returns ENCODED_PASSWORD

        userService = UserService(userRepository, passwordTokenRepository, emailService, passwordEncoder, authService)
    }

    @Test(expected = UserNotFoundException::class)
    fun `Should throw UserNotFoundException when no username or email in request`() {
        userService.createPasswordResetToken("", PasswordResetDto())
    }

    @Test(expected = UserNotFoundException::class)
    fun `Should throw UserNotFoundException when username does not exist`() {
        userService.createPasswordResetToken(
                SERVER_NAME,
                PasswordResetDto().apply { username = INVALID_USERNAME }
        )
    }

    @Test(expected = UserNotFoundException::class)
    fun `Should throw UserNotFoundException when email does not exist`() {
        userService.createPasswordResetToken(
                SERVER_NAME,
                PasswordResetDto().apply { email = INVALID_EMAIL }
        )
    }

    @Test
    fun `Should create password reset token for valid username`() {
        val passwordResetDto = PasswordResetDto().apply { username = VALID_USERNAME }

        every { passwordTokenRepository.save(any()) } returns PasswordResetToken()

        userService.createPasswordResetToken(SERVER_NAME, passwordResetDto)

        val savedToken = slot<PasswordResetToken>()

        verify { emailService.sendPasswordReset(any(), any(), SERVER_NAME) }
        verify { passwordTokenRepository.save(capture(savedToken)) }

        assertNotNull(savedToken.captured.user)
        assertNotNull(savedToken.captured.token)
    }

    @Test
    fun `Should create password reset token for valid email address`() {
        val passwordResetDto = PasswordResetDto().apply { email = VALID_EMAIL }

        every { passwordTokenRepository.save(any()) } returns PasswordResetToken()

        userService.createPasswordResetToken(SERVER_NAME, passwordResetDto)

        val savedToken = slot<PasswordResetToken>()

        verify { emailService.sendPasswordReset(any(), any(), SERVER_NAME) }
        verify { passwordTokenRepository.save(capture(savedToken)) }

        assertNotNull(savedToken.captured.user)
        assertNotNull(savedToken.captured.token)
    }

    @Test(expected = InvalidPasswordTokenException::class)
    fun `Should throw invalid password token exception when token doesn't exist`() {
        userService.processPasswordResetToken(INVALID_TOKEN)
    }

    @Test(expected = InvalidPasswordTokenException::class)
    fun `Should throw invalid password token exception when token has expired`() {
        userService.processPasswordResetToken(EXPIRED_TOKEN)
    }

    @Test
    fun `Should enable password reset for valid token`() {
        userService.processPasswordResetToken(VALID_TOKEN)

        val processedUserDetails = slot<RecipeUserDetails>()

        verify { authService.authenticateUser(capture(processedUserDetails), Role.CHANGE_PASSWORD) }

        assertEquals(VALID_USERNAME, processedUserDetails.captured.username)
        assertTrue(processedUserDetails.captured.changePasswordAccess)
    }

    @Test
    fun `Should change password successfully`() {
        val user = User()

        every { userRepository.save(user) } returns user

        userService.changePassword(user, PASSWORD)

        val savedUser = slot<User>()

        verify { userRepository.save(capture(savedUser)) }
        verify { passwordTokenRepository.deleteByUser(user) }
        verify { authService.disablePasswordReset() }

        assertEquals(ENCODED_PASSWORD, savedUser.captured.password)
    }

    @Test(expected = UserNotFoundException::class)
    fun `Should throw UserNotFoundException when user does not exist`() {
        userService.getUser(INVALID_USERNAME)
    }

    @Test
    fun `Should get user when valid username provided`() {
        val user = User()

        every { userRepository.findByUsername(VALID_USERNAME) } returns user

        assertEquals(user, userService.getUser(VALID_USERNAME))
    }

    @Test
    fun `Should delete account successfully`() {
        val user = User()

        every { authService.loggedInUser } returns user

        userService.deleteAccount()

        verify { userRepository.delete(user) }
    }

    @Test
    fun `Should turn off instructions if the user is new`() {
        val user = User().apply { newUser = true }

        every { authService.loggedInUser } returns user
        every { userRepository.save(user) } returns user

        userService.turnOffInstructions()

        val submittedUser = slot<User>()

        verify { userRepository.save(capture(submittedUser)) }

        assertFalse(submittedUser.captured.newUser)
    }

    @Test
    fun `Should not turn off instructions if the user is new`() {
        val user = User().apply { newUser = false }

        every { authService.loggedInUser } returns user

        userService.turnOffInstructions()

        verify(exactly = 0) { userRepository.save(user) }
    }
}